### DISCUSSION TITLE
Nested Eloquent Collection Hydration for non existing model and relation.

### DISCUSSION THREAD

**Discussion by juliomotol** [livewire/livewire#4140] [7. Bugs (archived)] [answered]
Livewire has a problem hydrating model relationship where in the model and its relation doesn't "exist" in the database. Given the following and assuming __*no*__ `Post` was given:

```php
class PostComponent extends Component
{
    public Post $post;

    public $rules = [
        'post.title' => ['required'],
        'post.content' => ['required'],
        // 'post.tags' => ['required', 'min:1'], // this is related with https://github.com/livewire/livewire/discussions/4107
        'post.tags.*.name' => ['required'],
    ];

    public function mount(Post $post = null)
    {
        // We'll initialize a `Post` model and its `tags` relation when its none was given.
        if (!$post->exists) {
            $this->post = new Post();
            $this->post->setRelation('tag', new EloquentCollection());
        }
    }

    public function addTag()
    {
        /**
         * At this point, the `tags` is hydrated as an EloquentCollection under the $post->attribute['tags']
         * 
         * But it should really be under $post->relations['tags]
         */
        $this->post->tags->push(new Tag());
    }
}
```

It will work fine on the first `addTag()` call. But since I'll be having a `wire:model="post.tags.{{ $key }}.name"`, triggering that will throw `Indirect modification of overloaded element` since as said earlier, its not hydrating the model's relation properly.

![image](https://user-images.githubusercontent.com/21353103/139774681-23abbfdb-23d1-4f4d-adc4-252292621a8f.png)

For now, we have gone with the ff  workaround:

```php
class PostsComponent extends Component
{
    ....
    public function hydratePost($value)
    {
        $tags = $this->post->tags->map(fn ($attributes) => new Tag($attributes));

        unset($this->post['tags']);

        $this->post->setRelation('tags', $tags);
    }
}
```

I've already done some digging around the other day and have found the potential cause.
https://github.com/livewire/livewire/blob/54864194f05daa62837ec9200e7c3cbbeb90b3ce/src/HydrationMiddleware/HydratePublicProperties.php#L202-L204

During dehydration, it's relations wasn't defined in the `memo.dataMeta` since the model was doesn't exists in the database, its only giving its class.

> Update 2021-11-02: Added exception screentshot

**Comment by joshhanley** [comment]
@juliomotol seems you have a mis-match in your code.

For your public property and rules, you are using `posts`(plural), but in your `mount` and `addTag` methods you are using `post` (singular).

Might be worth checking that out first to see if that is causing any issues.

Hope this helps!

**Reply by juliomotol** [reply]
My bad, those were typos

**Reply by joshhanley** [reply]
No worries!

**Comment by juliomotol** [comment]
More information dropping in...

After the component was `booted()` on the initial request, the model looks like so:
![image](https://user-images.githubusercontent.com/21353103/139775375-d5923aa9-09f0-47e0-9e90-ddad96da0a49.png)

Upon triggering the `wire:model`, its sends the following in its `dataMeta`

```json
{
    "serverMemo": {
        "data": {
            "returnReport": {
                "sap_merchant_branch_id": 1,
                "date": "2021-11-02T00:00:00.000000Z",
                "type": "bad-order",
                "details": [
                    { "sap_product_id": 9, "quantity": 0 },
                    { "sap_product_id": 18, "quantity": 0 },
                    { "sap_product_id": 6, "quantity": 0 },
                    ...
                ]
            }
        },
        "dataMeta": {
            "models": {
                "returnReport": { "class": "App\\Models\\ReturnReport\\ReturnReport" } // There's no relation defined in here even though we had one during its initial mounting...
            }
        }
    }
}
```
and the `$returnReport` now looks like...
![image](https://user-images.githubusercontent.com/21353103/139776188-36b896a9-f67a-4beb-b7c3-4b213d1b87f0.png)

Notice that:
- The `details` is still an instance of EloquentCollection but its items are not mapped into the Model. They're still an array in this case.
- The `details` is both in the `$model->attributes` and `$model->relations` and has the same reference of `#2974`.
- When doing `$model->details[0]->quantity` it will prioritize the `details` in `$model->attributes` instead of `$model->relations`

**Comment by joshhanley** [comment]
@juliomotol I had a quick look at this, and I'm not sure what we can do to fix it. 

We use Laravel's queue serialiser to transform a model and its relations into an array that can be sent to the front end. But I don't think it supports serialising and deserialising new models, as it looks like it is not recognising when there is a relation added to the model that hasn't been loaded from the database.

As such it's not passing the correct metadata on to the front end, which in turn means it's not available on the backend when the model it rehydrated.

I probably wouldn't use the deep nesting like that. If you need to manage a collection with new models in it, just make the collection a property on your component and assign to it from your original relation, and assign back on save. `$this->tags = $this->post->tags;`

If you are interested, it happens here:

https://github.com/livewire/livewire/blob/54864194f05daa62837ec9200e7c3cbbeb90b3ce/src/HydrationMiddleware/HydratePublicProperties.php#L202-L204

Hope this helps!

**Comment by juliomotol** [comment, marked as answer]
Bummer. I'd like to raise this issue to the `laravel/framework` itself but I don't see them consuming this fix. We'll just continue with our workaround.

```php
class PostsComponent extends Component
{
    ....
    public function hydratePost($value)
    {
        if ($this->post->exists) {
            return;
        }

        $tags = $this->post->tags->map(fn ($attributes) => new Tag($attributes));

        unset($this->post['tags']);

        $this->post->setRelation('tags', $tags);
    }
}
```

> I probably wouldn't use the deep nesting like that. If you need to manage a collection with new models in it, just make the collection a property on your component and assign to it from your original relation, and assign back on save. `$this->tags = $this->post->tags;`

Yeah, previously that is how we used to do things, but IMO, its much cleaner this way. Specially since nested model hydration is implemented in v2.6, we'd prefer to have the same behavior even it takes some added steps to it.

**Reply by joshhanley** [reply]
@juliomotol yeah they probably won't.

Yeah I understand it would definitely be cleaner!
